walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,96 @@
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
+  
+  private Collection<TimeRange> timeOptions = new ArrayList<TimeRange>();
+  
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    timeOptions.clear();
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+    
+    if (request.getAttendees().isEmpty() || events.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    
+    Set<String> allEventAttendees = new HashSet<>();
+    for (Event event : events) {
+      allEventAttendees.addAll(event.getAttendees());
+    }
+    Set<String> allRequestAttendees = new HashSet<>();
+    allRequestAttendees.addAll(request.getAttendees());
+    for (String attendee : allRequestAttendees) {
+      if (!allEventAttendees.contains(attendee)) {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+      }
+    }
+    
+    List<Event> eventsList = new ArrayList<Event>();
+    eventsList.addAll(events);
+    Collections.sort(eventsList, Event.ORDER_BY_START);
+    TimeRange firstEventTime = eventsList.get(0).getWhen();
+    TimeRange lastEventTime = eventsList.get(eventsList.size() - 1).getWhen();
+    boolean earlyEnd = false;
+    if (TimeRange.START_OF_DAY != firstEventTime.start()) {
+      addTimeRange(TimeRange.START_OF_DAY, firstEventTime.start());
+    }
+    for (int i = 1; i < eventsList.size(); i++) {
+      TimeRange currentEventTime = eventsList.get(i - 1).getWhen();
+      TimeRange nextEventTime = eventsList.get(i).getWhen();
+      TimeRange skipNextTime = null; 
+      if (i + 1 < eventsList.size()) {
+        skipNextTime = eventsList.get(i + 1).getWhen();
+      }
+      if (!backToBack(currentEventTime, nextEventTime)) {
+        if (!currentEventTime.overlaps(nextEventTime)) {
+          addTimeRange(currentEventTime.end(), nextEventTime.start());
+        }
+        if (currentEventTime.contains(nextEventTime)) {
+          if (i + 1 < eventsList.size()) {
+            addTimeRange(currentEventTime.end(), skipNextTime.start());
+          } else {
+            earlyEnd = true;
+            if (lastEventTime.end() - 1 != TimeRange.END_OF_DAY) {
+              addTimeRange(currentEventTime.end(), TimeRange.END_OF_DAY);
+            }
+          }
+        } 
+      } 
+    }
+    if ((lastEventTime.end() - 1 != TimeRange.END_OF_DAY) && !earlyEnd) {
+      addTimeRange(lastEventTime.end(), TimeRange.END_OF_DAY);
+    }
+    
+    for (TimeRange meetingOption : timeOptions) {
+      if (meetingOption.duration() >= request.getDuration()) {
+        return timeOptions;
+      }
+    }
+    
+    return Arrays.asList();
+  }
+  
+  private void addTimeRange(int start, int end) {
```

Comment:
This needs either a better, domain specific name, or a comment. It's hard to build an intuition about this from call sites.

"addTimeOption" would at least tie this to the timeOptions member variable.

Reply:
Done